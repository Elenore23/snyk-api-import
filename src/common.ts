export const IMPORT_LOG_NAME = 'imported-targets.log';
export const FAILED_LOG_NAME = 'failed-imports.log';
export const FAILED_PROJECTS_LOG_NAME = 'failed-projects.log';
export const FAILED_POLLS_LOG_NAME = 'failed-polls.log';
export const IMPORT_JOBS_LOG_NAME = 'import-jobs.log';
export const IMPORTED_PROJECTS_LOG_NAME = 'imported-projects.log';
export const IMPORTED_BATCHES_LOG_NAME = 'imported-batches.log';
export const IMPORT_JOB_RESULTS = 'import-job-results.json';
