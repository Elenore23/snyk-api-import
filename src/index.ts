#!/usr/bin/env node

import * as yargs from 'yargs';
export * from './lib';

yargs
  .commandDir('cmds')
  .demandCommand()
  .help()
  .argv
