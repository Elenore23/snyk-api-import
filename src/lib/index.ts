import 'source-map-support/register';
import * as needle from 'needle';
import * as sleep from 'sleep-promise';
import { Url } from 'url';

interface Target {
  name?: string; // Gitlab, GitHub, GH Enterprise, Bitbucket Cloud and Azure Repos, Bitbucket Server, Azure Container Registry, Elastic Container Registry, Artifactory Container Registry, Docker Hub
  appId?: string; // Heroku, CloudFoundry, Pivotal & IBM Cloud
  functionId?: string; // AWS Labmda
  slugId?: string; // Heroku
  projectKey?: string; // Bitbucket Server
  repoSlug?: string; // Bitbucket Server
  id?: number; // Gitlab
  owner: string; // Gitlab, GitHub, GH Enterprise, Bitbucket Cloud and Azure Repos
  branch: string; // Gitlab, GitHub, GH Enterprise, Bitbucket Cloud and Azure Repos
}

interface FilePath {
  path: string;
}
const SNYK_HOST = 'https://dev.snyk.io';
const MIN_RETRY_WAIT_TIME = 30000;
const MAX_RETRY_COUNT = 1000;

export async function importTarget(
  orgId: string,
  integrationId: string,
  target: Target,
  files?: FilePath[],
): Promise<{ pollingUrl: string }> {
  const apiToken = process.env.SNYK_API_TOKEN;
  if (!apiToken) {
    throw new Error(
      `Please set the SNYK_API_TOKEN e.g. export SNYK_API_TOKEN='*****'`,
    );
  }
  console.log(
    'Importing target:',
    { orgId, integrationId, target },
    Object.keys(target).length,
  );

  if (!orgId || !integrationId || Object.keys(target).length === 0) {
    throw new Error(
      `Missing required parameters. Please ensure you have set: orgId, integrationId, target.
      \nFor more information see: https://snyk.docs.apiary.io/#reference/integrations/import-projects/import`,
    );
  }
  try {
    const body = {
      target,
      files,
    };

    const res = await needle(
      'post',
      `${SNYK_HOST}/api/v1/org/${orgId}/integrations/${integrationId}/import`,
      body,
      {
        json: true,
        // eslint-disable-next-line @typescript-eslint/camelcase
        read_timeout: 30000,
        headers: {
          Authorization: `token ${apiToken}`,
        },
      },
    );
    const pollingUrl = res.headers['location'];
    if (!pollingUrl) {
      throw new Error(
        'No import location url returned. Please re-try the import.',
      );
    }
    console.log('** res', res.body, pollingUrl);
    return { pollingUrl };
  } catch (error) {
    const err: {
      message?: string | undefined;
      innerError?: string;
    } = new Error('Could not complete API import');
    err.innerError = error;
    throw err;
  }
}

interface Status {
  status: 'pending' | 'failed' | 'complete';
}

interface Project {
  targetFile?: string;
  success: boolean;
  projectUrl: Url;
}

interface Log {
  name: 'org2/repo2';
  created: '2018-07-23T15:21:10.644Z';
  status: Status;
  projects: Project[];
}

interface PollImportResponse {
  id: string;
  status: 'pending' | 'failed' | 'complete';
  created: string;
  logs: Log[];
}

export async function pollImportUrl(
  locationUrl: string,
  retryCount = MAX_RETRY_COUNT,
  retryWaitTime = MIN_RETRY_WAIT_TIME,
): Promise<object> {
  const apiToken = process.env.SNYK_API_TOKEN;
  // const apiToken = process.env.SNYK_API_TOKEN_LOCAL;

  if (!apiToken) {
    throw new Error(
      `Please set the SNYK_API_TOKEN e.g. export SNYK_API_TOKEN='*****'`,
    );
  }

  if (!locationUrl) {
    throw new Error(
      `Missing required parameters. Please ensure you have provided: location url.
      \nFor more information see: https://snyk.docs.apiary.io/#reference/integrations/import-projects/import`,
    );
  }
  try {
    const res = await needle('get', `${locationUrl}`, {
      json: true,
      // eslint-disable-next-line @typescript-eslint/camelcase
      read_timeout: 30000,
      headers: {
        Authorization: `token ${apiToken}`,
      },
    });
    const importStatus: PollImportResponse = res.body;
    console.log(`Import task status is |${importStatus.status}|`);
    if (
      importStatus.status &&
      importStatus.status !== 'complete' &&
      retryCount > 0
    ) {
      await sleep(retryWaitTime);
      const increasedRetryWaitTime =
        retryWaitTime + retryWaitTime * 0.1 * (MAX_RETRY_COUNT - retryCount);
        console.log(`Will re-check import task in |${increasedRetryWaitTime}| ms`);
      return await pollImportUrl(
        locationUrl,
        --retryCount,
        increasedRetryWaitTime,
      );
    }
    return res.body;
  } catch (error) {
    console.log('Could not complete API import:', error);
    const err: {
      message?: string | undefined;
      innerError?: string;
    } = new Error('Could not complete API import');
    err.innerError = error;
    throw err;
  }
}
