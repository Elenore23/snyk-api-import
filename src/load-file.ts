import * as fs from 'fs';
import * as path from 'path';
import * as debugLib from 'debug';
const debug = debugLib('snyk:load-file');

export async function loadFile(name: string): Promise<string> {
  const filename = path.resolve(name);
  try {
    return await fs.readFileSync(filename, 'utf8');
  } catch (error) {
    debug(error.message);
    throw new Error(`File can not be found at location: ${filename}`);
  }
}
