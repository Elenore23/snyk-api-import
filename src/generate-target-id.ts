import * as _ from 'lodash';

import { targetProps } from './common';
import { Target } from './lib/types';

export function generateTargetId(
  orgId: string,
  integrationId: string,
  target: Target,
): string {
  const targetData =  _.pick(target, ...targetProps)
  return `${orgId}:${integrationId}:${Object.values(targetData).join(':')}`;
}
