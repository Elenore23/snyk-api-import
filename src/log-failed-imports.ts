import * as fs from 'fs';
import * as debugLib from 'debug';

import { Target } from './lib/types';
import { getLoggingPath } from './lib/get-logging-path';
import { FAILED_LOG_NAME } from './common';

const debug = debugLib('snyk:import-projects-script');

export async function logFailedImports(
  orgId: string,
  integrationId: string,
  target: Target,
  loggingPath: string = getLoggingPath(),
  locationUrl?: string,
): Promise<void> {
  try {
    const log = `${integrationId}:${Object.values(target).join(
      ':',
    )}:${locationUrl},`;
    fs.appendFileSync(`${loggingPath}/${orgId}.${FAILED_LOG_NAME}`, log);
  } catch (e) {
    debug('Failed to log failed imports at location: ', loggingPath);
    // do nothing
  }
}
