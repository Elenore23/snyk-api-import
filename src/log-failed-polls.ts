import * as bunyan from 'bunyan';

import { getLoggingPath } from './lib/get-logging-path';
import { FAILED_POLLS_LOG_NAME } from './common';

export async function logFailedPollUrls(
  locationUrl: string,
  errorData: {
    errorMessage: string,
    [name: string]: any;
  },
  loggingPath: string = getLoggingPath(),
): Promise<void> {
  try {
    const orgId = locationUrl.split('/').slice(-5)[0];
    const log = bunyan.createLogger({
      name: 'snyk:import-projects-script',
      level: 'error',
      streams: [{
        level: 'error',
        path: `${loggingPath}/${orgId}.${FAILED_POLLS_LOG_NAME}`,
      }],
    });
    log.error({ orgId, locationUrl, ...errorData }, 'Failed to poll url');
  } catch (e) {
    // do nothing
  }
}
