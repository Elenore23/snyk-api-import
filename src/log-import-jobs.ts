import * as fs from 'fs';
import { getLoggingPath } from './lib/get-logging-path';
import { IMPORT_JOBS_LOG_NAME } from './common';

export async function logImportJobsPerOrg(
  orgId: string,
  pollingUrl: string,
  loggingPath: string = getLoggingPath(),
): Promise<void> {
  try {
    fs.appendFileSync(`${loggingPath}/${orgId}.${IMPORT_JOBS_LOG_NAME}`, `${pollingUrl},`);
  } catch (e) {
    // do nothing
  }
}
