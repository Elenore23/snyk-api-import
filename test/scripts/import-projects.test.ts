import * as path from 'path';
import * as fs from 'fs';
import { ImportProjects } from '../../src/scripts/import-projects';
import { IMPORT_PROJECTS_FILE_NAME, IMPORT_LOG_NAME } from '../../src/common';

describe('Import projects script', () => {
  const logPath = path.resolve(__dirname, IMPORT_LOG_NAME);
  afterEach(() => {
    fs.unlinkSync(logPath);
  });
  it.only('succeeds to import targets from file', async () => {
    const projects = await ImportProjects(
      path.resolve(__dirname + `/fixtures/${IMPORT_PROJECTS_FILE_NAME}`),
      __dirname,
    );
    expect(projects).not.toBe([]);
    expect(projects[0]).toMatchObject({
      projectUrl: expect.any(String),
      success: true,
      targetFile: expect.any(String),
    });
    const logFile = fs.readFileSync(logPath, 'utf8');
    expect(logFile).toMatch('shallow-goof-policy');
    expect(logFile).toMatch('composer-with-vulns');
    expect(logFile).toMatch('ruby-with-versions:');
  }, 30000000);
});

describe('Import skips previously imported', () => {
  const logPath = path.resolve(
    __dirname + '/fixtures/with-import-log',
    IMPORT_LOG_NAME,
  );
  it('succeeds to import targets from file', async () => {
    const projects = await ImportProjects(
      path.resolve(
        __dirname + `/fixtures/with-import-log/${IMPORT_PROJECTS_FILE_NAME}`,
      ),
      __dirname + '/fixtures/with-import-log',
    );
    expect(projects.length === 0).toBeTruthy();
    const logFile = fs.readFileSync(logPath, 'utf8');
    expect(logFile).toMatchSnapshot();
  }, 30000000);
});

describe('Skips issues', () => {
  const logPath = path.resolve(__dirname, IMPORT_LOG_NAME);
  it('Skips any badly formatted targets', async () => {
    // TODO: ensure all failures are logged & assert it is present in the log
    const projects = await ImportProjects(
      path.resolve(__dirname + '/fixtures/import-projects-invalid-target.json'),
      __dirname,
    );
    expect(projects.length === 0).toBeTruthy();
    let logFile = null;
    try {
      logFile = fs.readFileSync(logPath, 'utf8');
    } catch (e) {
      expect(logFile).toBeNull();
    }
  }, 300);
});

describe('Error handling', () => {
  it('shows correct error when input can not be loaded', async () => {
    expect(
      ImportProjects(`do-not-exist/${IMPORT_PROJECTS_FILE_NAME}`),
    ).rejects.toThrow('File can not be found at location');
  }, 300);
  it('shows correct error when input is invalid json', async () => {
    expect(
      ImportProjects(
        path.resolve(__dirname + '/fixtures/import-projects-invalid.json'),
      ),
    ).rejects.toThrow('Failed to parse targets from');
  }, 300);
});
