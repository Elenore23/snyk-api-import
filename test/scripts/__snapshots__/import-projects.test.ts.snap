// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import skips previously imported succeeds to import targets from file 1`] = `
"f0125d9b-271a-4b50-ad23-80e12575a1bf:c4de291b-e083-4c43-a72c-113463e0d268:composer-with-vulns:snyk-fixtures:master,
"
`;
